### PR TITLE
[Backport][v1.6.x] Fix the wrong result of sum, mean, argmin, argmax when inputs contain inf or nan

### PR DESCRIPTION
**Pull request by wkcn**
Hi, there.
In v1.6.x, there is a bug of reduce operators when the inputs contain inf and nan.
I cherry-pick 52716dee into the branch v1.6.x.

Related PR: https://github.com/apache/incubator-mxnet/pull/16234

### CODE REVIEW

File: src/operator/tensor/elemwise_unary_op.h [hunk 1]
```diff
@@ -660,6 +660,134 @@ void AroundOpForward(const nnvm::NodeAttrs& attrs,
...
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int i,
+                                  DType* out_data,
+                                  const DType* in_data,
+                                  const DType nan,
+                                  const DType posinf,
+                                  const DType neginf) {
+    DType val = in_data[i];
+    if (mshadow_op::IsNan<DType>(val))  val = nan;
+    if (val > 0 && mshadow_op::IsInf(val))  val = posinf;
```
Comment (wkcn) [inline src/operator/tensor/elemwise_unary_op.h]: Do I need to delete the modification in src/operator/tensor/elemwise_unary_op.h?
The class NumpyNanToNumParam was written in other PR, but I only modify several lines in it.

**Comment by ptrendx** [comment]
@wkcn is that class part of 1.6? If not, then I agree you should delete those modifications - we do not want to introduce new functionality into 1.6 at this point, just fixes.

**Comment by wkcn** [comment]
Hi @ptrendx , I have deleted these modifications : )

File: 3rdparty/mshadow/mshadow/base.h [hunk 2]
```diff
@@ -606,6 +606,64 @@ struct divto {
...
+using std::isnan;
+using std::isinf;
+#endif
+
+/*! \brief
+ *  determines if the given floating point
+ *  number is not a number */
+namespace isnan_typed {
+  template<typename DType>
+  MSHADOW_XINLINE bool IsNan(volatile DType val) {
```
Comment (larroy) [inline 3rdparty/mshadow/mshadow/base.h]: what's the point of volatile here?

**Pull request merged**